src/ipc/ipc4/helper.c

```diff
@@ -182,18 +182,39 @@ static int ipc_pipeline_module_free(uint32_t pipeline_id)
 
 	icd = ipc_get_comp_by_ppl_id(ipc, COMP_TYPE_COMPONENT, pipeline_id);
 	while (icd) {
-		struct comp_buffer *sink;
-		struct list_item *sink_list;
-		struct list_item *tmp_list;
+		struct list_item *list, *_list;
+		struct comp_buffer *buffer;
 
 		/* free sink buffer allocated by current component in bind function */
-		list_for_item_safe(sink_list, tmp_list, &icd->cd->bsink_list) {
-			sink = container_of(sink_list, struct comp_buffer, source_list);
+		list_for_item_safe(list, _list, &icd->cd->bsink_list) {
+			struct comp_buffer __sparse_cache *buffer_c;
+			struct comp_dev *sink;
+
+			buffer = container_of(list, struct comp_buffer, source_list);
+			pipeline_disconnect(icd->cd, buffer, PPL_CONN_DIR_COMP_TO_BUFFER);
+			buffer_c = buffer_acquire(buffer);
+			sink = buffer_c->sink;
+			buffer_release(buffer_c);
+
+			/* free the buffer only when the sink module has also been disconnected */
+			if (!sink)
+				buffer_free(buffer);
+		}
+
+		/* free source buffer allocated by current component in bind function */
+		list_for_item_safe(list, _list, &icd->cd->bsource_list) {
```

Comment:
each component free its sink buffer then all the buffer will be free.  If a component free both sink and source buffer then the connected component will have no sink buffer or source buffer to free

Reply:
@RanderWang but the code frees both buffers only if the buffer isnt connected to a source and a sink and then it would be correct.